src/libraries/Common/src/SourceGenerators/ImmutableEquatableArray.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics.Hashing;
+
+namespace SourceGenerators
+{
+    /// <summary>
+    /// Provides an immutable list implementation which implements sequence equality.
+    /// </summary>
+    public sealed class ImmutableEquatableArray<T> : IEquatable<ImmutableEquatableArray<T>>, IReadOnlyList<T>
```

Comment:
Is there a reason this is a class, instead of a struct?

Reply:
I wanted to avoid having to deal with `default` instances which would also make equality comparison more complicated. Plus, it's more convenient when your model needs to have nullable lists.